src/board.rs

```diff
@@ -146,83 +157,264 @@ impl Chessboard {
-//     // If it's a capture, clear the bit for the captured piece
-//     // Update en passant square and castling rights accordingly
-//     // Switch turns: white_turn = !white_turn;
-// }
+        // Parse castling rights
+        let fen_castle = fen_parts[2];
+        let mut castles = 0;
+        for c in fen_castle.chars() {
+            let v = match c {
+                'K' => 0b1000,
+                'Q' => 0b0100,
+                'k' => 0b0010,
+                'q' => 0b0001,
+                _ => 0b0,
+            };
+            castles |= v;
+        }
+        chessboard.castling_rights = castles;
 
-// generate_legal_moves() {
-//     legal_moves = []
-
-//     for each piece in pieces_of_current_turn {
-//         possible_moves = generate_moves_for_piece(piece)
-//         for each move in possible_moves {
-//             if is_legal_move(move) {
-//                 legal_moves.push(move)
-//             }
-//         }
-//     }
-//
-//     return legal_moves
-// }
+        // Parse en passant square
+        let fen_passant = fen_parts[3];
+        if fen_passant != "-" {
+            if let (Some(col), Some(row)) = (
+                fen_passant.chars().nth(0).map(|c| c.to_ascii_uppercase()),
+                fen_passant.chars().nth(1).and_then(|c| c.to_digit(10)),
+            ) {
+                if (1..=8).contains(&row) {
+                    let col_value: u8 = match col {
+                        'A' => 1,
+                        'B' => 2,
+                        'C' => 3,
+                        'D' => 4,
+                        'E' => 5,
+                        'F' => 6,
+                        'G' => 7,
+                        'H' => 8,
+                        _ => 0, // Handle unexpected characters
+                    };
+                    chessboard.en_passant = col_value + 8 * (row as u8 - 1);
+                }
+            }
+        }
 
-// generate_moves_for_piece(piece) {
-//     // Generate all possible moves for the given piece
-//     // Consider piece-specific movement rules (e.g., pawn's initial double move, castling for king, etc.)
-// }
+        // Ignore the rest of the FEN string for now
+        return chessboard;
+    }
+
+    // Serializer function that converts a Chessboard struct to a FEN (Forsyth–Edwards Notation) string
+    pub fn to_string(&self) -> String {
+        // Initialize a vector to store FEN components as strings
+        let mut string_array: Vec<String> = Vec::with_capacity(6);
+        let mut fen_string;
+
+        // Piece placement
+        for rank in (1..=8).rev() {
+            let mut empty_squares = 0;
+            let mut row_string = String::new();
+
+            // Iterate through each file (column) in the rank
+            for file in 1..=8 {
+                let square_index = (rank - 1) * 8 + (file - 1);
+
+                // Determine the piece on the current square
+                let piece = if (self.white_pawns >> square_index) & 1 != 0 {
```

Comment:
Surely, there is a way to put this into a for loop? I'm too tired to think of it tho

Reply:
Lmk when you think of something!

Reply:
https://github.com/chanson02/KnO3/commit/928dbc211faf96c28867efdab2bfcc20d763b8c2